accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -101,134 +143,119 @@ impl<K: Ord + Copy, T> DeltaDataCollection<K, T> {
+            match combine_delta_data(&u1.0, u2.0).map(|d| d.map(|d| DataDeltaUndo(d)))? {
+                Some(d) => Ok(DeltaMapOp::Insert(DeltaMapElement::DeltaUndo(d))),
+                None => Ok(DeltaMapOp::Erase),
+            }
         }
     }
 }
 
 /// Given two deltas, combine them into one delta, this is the basic delta data composability function
-fn combine_delta_data<T>(
+fn combine_delta_data<T: Clone + PartialEq>(
     lhs: &DataDelta<T>,
     rhs: DataDelta<T>,
-) -> Result<DeltaMapOp<DataDelta<T>>, Error> {
+) -> Result<Option<DataDelta<T>>, Error> {
     match (lhs, rhs) {
         (DataDelta::Create(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
-        (DataDelta::Create(_), DataDelta::Modify(d)) => Ok(DeltaMapOp::Write(DataDelta::Create(d))),
-        (DataDelta::Create(_), DataDelta::Delete) => {
-            // if lhs had a creation, and we delete, this means nothing is left and there's a net zero to return
-            Ok(DeltaMapOp::Delete)
+        (DataDelta::Create(lhs), DataDelta::Modify(from, to)) => {
+            ensure!(lhs == &from, Error::DeltaDataMismatch);
+            Ok(Some(DataDelta::Create(to)))
+        }
+        (DataDelta::Create(lhs), DataDelta::Delete(rhs)) => {
+            // if lhs had a creation, and we DataDeltaUndo::Delete, this means nothing is left and there's a net zero to return
+            ensure!(lhs == &rhs, Error::DeltaDataMismatch);
+            Ok(None)
+        }
+        (DataDelta::Modify(_, _), DataDelta::Create(_)) => {
+            Err(Error::DeltaDataCreatedMultipleTimes)
+        }
+        (DataDelta::Modify(lhs1, lhs2), DataDelta::Modify(rhs1, rhs2)) => {
+            if lhs1 == &rhs2 && lhs2 == &rhs1 {
+                Ok(None)
+            } else {
+                Ok(Some(DataDelta::Modify(lhs1.clone(), rhs2)))
+            }
         }
-        (DataDelta::Modify(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
-        (DataDelta::Modify(_), DataDelta::Modify(d)) => Ok(DeltaMapOp::Write(DataDelta::Modify(d))),
-        (DataDelta::Modify(_), DataDelta::Delete) => {
-            // if lhs had a modification, and we delete, this means nothing is left and there's a net zero to return
-            Ok(DeltaMapOp::Delete)
+        (DataDelta::Modify(from, to), DataDelta::Delete(rhs)) => {
+            ensure!(to == &rhs, Error::DeltaDataMismatch);
+            Ok(Some(DataDelta::Delete(from.clone())))
         }
-        (DataDelta::Delete, DataDelta::Create(d)) => Ok(DeltaMapOp::Write(DataDelta::Create(d))),
-        (DataDelta::Delete, DataDelta::Modify(_)) => Err(Error::DeltaDataModifyAfterDelete),
-        (DataDelta::Delete, DataDelta::Delete) => Err(Error::DeltaDataDeletedMultipleTimes),
+        (DataDelta::Delete(lhs), DataDelta::Create(rhs)) => {
+            if lhs == &rhs {
+                Ok(None)
+            } else {
+                Ok(Some(DataDelta::Modify(lhs.clone(), rhs)))
+            }
+        }
```

Comment:
I think there's a subtle issue here.  `None` is not the same as `Some(DeltaData::Modify(x, x))`. The former succeeds unconditionally. The latter errors out if the original value is not `x` functionally being equivalent to and assertion that the value is equal to `x`.

Imho this should be `Modify(lhs, rhs)` even if `lhs` and `rhs` are equal. That way we get the same result applying the deltas one by one as we do combining the deltas into one and applying it.

Similar issue is present in the `Create`/`Delete` pair. The `None` result is returned there which succeeds unconditionally. However, the individual operations only succeed if the value is initially not present (`None`).

In addition to that, the comment about cloning above also applies here.

Reply:
Interesting. I agree that producing a delta in all the cases can give us additional protection when merging the result with data. Or at least for the case with Modify.

But this raises a question of what the result of Create + Delete might be. Delete? 
Consider the cases:
```
Create + Delete = Delete
None + Create + Delete = Some + Delete = None
None + (Create + Delete) = None + Delete = Error
Some + Create + Delete = Error + Delete = Error
Some + (Create + Delete) = Some + Delete = None
```
or if it's a No-op:
```
Create + Delete = No-op
None + Create + Delete = Some + Delete = None
None + (Create + Delete) = None + No-op = None
Some + Create + Delete = Error + Delete = Error
Some + (Create + Delete) = Some + No-op = Some
```
The second variant is closer to the truth because it gives None for the 2nd case. But both of them don't produce an error for the 4th case, which unfortunately breaks our rule of associativity.

Reply:
> But this raises a question of what the result of Create + Delete might be.

I think to make `Create` + `Delete` equivalent to applying `Create` and  then applying `Delete` one by one, we need a delta that fails when an element is already there and does nothing otherwise. I.e. another case like `LeaveDeleted` is needed to handle it (in addition to `Create`, `Delete`, `Modify`).

I briefly mention the issue in the long comment I just posted. There a different representation is used where only `Modify` operation is there and missing values are represented by using `Option<_>` as `T`. Then it becomes `Modify(None, None)`